# Introduction

This project focuses on analyzing the **Spaceship Titanic Dataset** to develop a *machine learning model* that can predict which passengers were transported to an alternate dimension during a collision with a spacetime anomaly. Set in the year 2912, this futuristic scenario presents a unique challenge in the field of data science and interstellar travel safety.

The dataset contains various attributes about the passengers including:

* Demographics
   * Age
   * Name
* Travel Information
   * Home Planet
   * Destination
   * Cabin location
* Onboard Activities
   * RoomService
   * FoodCourt
   * ShoppingMall
   * Spa
   * VRDeck
* Special Statuses
   * VIP
   * CryoSleep

The target variable is whether the passenger was `Transported` to an alternate dimension.

As data scientists working with **Interstellar Rescue Operations**, our objective is to thoroughly explore this data to uncover patterns and insights that can inform the development of a robust predictive model. This will enable rescue crews to identify which passengers were transported and focus their retrieval efforts accordingly.

The analysis will progress through the following key steps:

1. **Exploratory Data Analysis** - Examining the distributions, ranges, and relationships between the features and target variable through statistical summaries and visualizations. Checking data quality and identifying any anomalies in the spaceship's records.

2. **Statistical Inference** - Formulating and testing hypotheses about transportation risk factors and quantifying uncertainty through confidence intervals.

3. **Machine Learning Modeling** - Applying a range of classification algorithms including random forests, gradient boosting, and potentially neural networks to predict transportation likelihood. Tuning hyperparameters and building ensembles to optimize predictive performance.

4. **Model Deployment** - Selecting the top performing model and preparing it for deployment to enable efficient identification of transported passengers in future space anomaly encounters.

Throughout this notebook, detailed commentary will be provided on the analytical approach, key findings, model results and ideas for further enhancement. The goal is to demonstrate a thoughtful, thorough analysis while documenting reproducible steps from data intake through model evaluation.

By accurately predicting which passengers were transported, this project aims to assist rescue operations in efficiently locating and retrieving lost travelers, potentially saving thousands of lives.

*Let's begin the analysis to uncover the mysteries hidden in the Spaceship Titanic's data.*

In [8]:
import sys
sys.path.append("../src")

In [10]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.impute import SimpleImputer
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.feature_selection import SelectKBest, f_classif
from xgboost import XGBClassifier
from lightgbm import LGBMClassifier
import optuna
from spaceship_titanic_utils import *

In [12]:
train_df = pd.read_csv("../data/train.csv")
test_df = pd.read_csv("../data/test.csv")

In [21]:
print("Train dataset first few rows:")
print(train_df.head())
print("\nTrain dataset shape:", train_df.shape)

Train dataset first few rows:
  PassengerId HomePlanet CryoSleep  Cabin  Destination   Age    VIP  \
0     0001_01     Europa     False  B/0/P  TRAPPIST-1e  39.0  False   
1     0002_01      Earth     False  F/0/S  TRAPPIST-1e  24.0  False   
2     0003_01     Europa     False  A/0/S  TRAPPIST-1e  58.0   True   
3     0003_02     Europa     False  A/0/S  TRAPPIST-1e  33.0  False   
4     0004_01      Earth     False  F/1/S  TRAPPIST-1e  16.0  False   

   RoomService  FoodCourt  ShoppingMall     Spa  VRDeck               Name  \
0          0.0        0.0           0.0     0.0     0.0    Maham Ofracculy   
1        109.0        9.0          25.0   549.0    44.0       Juanna Vines   
2         43.0     3576.0           0.0  6715.0    49.0      Altark Susent   
3          0.0     1283.0         371.0  3329.0   193.0       Solam Susent   
4        303.0       70.0         151.0   565.0     2.0  Willy Santantines   

   Transported  
0        False  
1         True  
2        False  
3     

In [16]:
print("\nTest dataset first few rows:")
print(test_df.head())
print("\nTest dataset shape:", test_df.shape)


Test dataset first few rows:
  PassengerId HomePlanet CryoSleep  Cabin  Destination   Age    VIP  \
0     0013_01      Earth      True  G/3/S  TRAPPIST-1e  27.0  False   
1     0018_01      Earth     False  F/4/S  TRAPPIST-1e  19.0  False   
2     0019_01     Europa      True  C/0/S  55 Cancri e  31.0  False   
3     0021_01     Europa     False  C/1/S  TRAPPIST-1e  38.0  False   
4     0023_01      Earth     False  F/5/S  TRAPPIST-1e  20.0  False   

   RoomService  FoodCourt  ShoppingMall     Spa  VRDeck              Name  
0          0.0        0.0           0.0     0.0     0.0   Nelly Carsoning  
1          0.0        9.0           0.0  2823.0     0.0    Lerome Peckers  
2          0.0        0.0           0.0     0.0     0.0   Sabih Unhearfus  
3          0.0     6652.0           0.0   181.0   585.0  Meratz Caltilter  
4         10.0        0.0         635.0     0.0     0.0   Brence Harperez  

Test dataset shape: (4277, 13)


The Spaceship Titanic dataset is divided into train and test sets. The train set contains **8,693 samples** with **14 features**, including the binary target variable 'Transported'. 

The test set has **4,277 samples** with **13 features**, excluding the target variable. Both sets share common features like 'PassengerId', 'HomePlanet', 'CryoSleep', 'Cabin', 'Destination', 'Age', 'VIP', and various amenity usage metrics such as 'RoomService' and 'FoodCourt'. 

The data includes a mix of numeric and categorical variables.